---
title: An R Markdown document converted from "/content/alignment and feature counting.ipynb"
output: html_document
---

### Import

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(c("Rsubread"))
```

```{r}
install.packages("here")
```

```{r}
library(here)
library(Rsubread)
```

## Introduction and data import

```{r}
filename <- here()
filename
```

```{r}
filename <- here("../output/Bam files (STAR)")
```

```{r}
fastq.files <- list.files(path = filename, pattern = ".fastq.gz$", full.names = TRUE)
fastq.files
```

## Quality control

```{r}
# Extract quality scores
for (q in fastq.files) {
    qs <- qualityScores(filename = q, nreads = 100)
    boxplot(qs, ylab = "Quality score", xlab = "Base position", main = substr(q,
        start = 17, stop = 35), cex = 0.5, col = "orange")
}
```

```{r}
# Check dimension of qs
dim(qs)
```

```{r}
# Check first few elements of qs with head
head(qs)
```

```{r}
boxplot(qs, ylab = "Quality score", xlab = "Base position", main = "SRR1552445.fastq", 
    cex = 0.5, col = "orange")
```

## Build the index

```{r}
buildindex(basename = "mm39", reference = "../reference/genome/mm39/mm39.fa")
# build index genome for mm39
```

```{r}
buildindex(basename = "hg38", reference = "../reference/genome/hg38/hg38.fa.gz")
# build index genome for hg38
```

```{r}
args(buildindex)
```

## Aligning reads to chromosomes of reference genome

```{r}
align(index = "../reference/genome_indexed/mm10/mm10", readfile1 = fastq.files,
    annot.inbuilt = "mm10")
# with mm10 and inbuild annotation
```

```{r}
align(index = "../reference/genome_indexed/mm10/mm10", readfile1 = fastq.files,
    annot.ext = "../reference/annotation/mm10/mm10.ncbiRefSeq.gtf",
    isGTF = TRUE)
# with mm10 and gtf annotation
```

```{r}
align(index = "../reference/genome_indexed/mm39/mm39", readfile1 = fastq.files,
    annot.ext = "../reference/annotation/mm39/mm39.ncbiRefSeq.gtf",
    isGTF = TRUE)
# with mm39 and gtf annotation
```

```{r}
align(index = "D:/RNA-seq Data/RNA-seq/reference/genome_indexed/hg38/hg38", readfile1 = fastq.files,
    annot.ext = "D:/RNA-seq Data/RNA-seq/reference/annotation/hg38/hg38.ncbiRefSeq.gtf",
    isGTF = TRUE)
# with mm39 and gtf annotation
```

```{r}
args(align)
```

```{r}
?align
```

```{r}
filename <- here("C:/Bam files (STAR)")
```

```{r}
bam.files <- list.files(path = filename, pattern = ".bam$", full.names = TRUE)
bam.files
length(bam.files)
```

```{r}
bam.files <- list.files(path = "./", pattern = ".BAM$", full.names = TRUE)
bam.files
```

```{r}
props <- propmapped(files = bam.files[15:19])
```

```{r}
props
```

```{r}
write.csv(props, "props.csv")
```

## Counting

```{r}
fc <- featureCounts(bam.files, annot.inbuilt = "mm10", , isPairedEnd = TRUE)
# with mm10 and inbuilt annotation
```

```{r}
fc <- featureCounts(bam.files, annot.ext = "D:/reference/annotation/mm10/mm10.ncbiRefSeq.gtf",
    isGTFAnnotationFile = TRUE)
# with mm10 and gtf annotation
```

```{r}
fc <- featureCounts(bam.files, annot.ext = "D:/reference/annotation/mm39/mm39.ncbiRefSeq.gtf",
    isGTFAnnotationFile = TRUE, isPairedEnd = TRUE)
# with mm39 and gtf annotation
```

```{r}
fc <- featureCounts(bam.files, annot.inbuilt = "hg38", isPairedEnd = TRUE)
# with hg38 and inbuilt annotation
```

```{r}
args(featureCounts)
```

```{r}
# See what slots are stored in fc
names(fc)
```

```{r}
# Take a look at the featurecounts stats
fc$stat
```

```{r}
# Take a look at the dimensions to see the number of genes
dim(fc$counts)
```

```{r}
## Take a look at the first 6 lines
head(fc$counts)
```

```{r}
head(fc$annotation)
```

```{r}
feature_stats <- as.data.frame(fc$stat)
write.csv(feature_stats, "feature_stats.csv")
```

```{r}
out_put <- as.data.frame(fc$counts)
write.csv(out_put, "counts.csv")
```

```{r}
# The TPM are giving in the table, but we can demonstrate the calculation in a
# few steps divide the number of reads for each transcript by it's length
# (reads per kilobase - RPK) sum the RPK values and divide by 1 million to get
# a scaling factor divide the RPK values by the scaling factor to get the TPM
rpk <- fc$counts/fc$annotation$Length
scale_factor <- sum(rpk)/1e+06
tpm <- rpk/scale_factor
rownames(tpm) <- rownames(fc$counts)
head(tpm)
```

```{r}
out_put <- as.data.frame(tpm)
write.csv(out_put, "counts_tpm.csv")
```

