---
title: An R Markdown document converted from "C:/Users/shaoq/OneDrive/PythonScripts/Tools/Analyzing
  RNA-seq data in the dorsal root ganglion neurons with limma, Glimma and edgeR.ipynb"
output: html_document
---

# Analyzing RNA-seq data in the dorsal root ganglion neurons with limma, Glimma and edgeR

##  Abstract
The ability to easily and efficiently analyse RNA-sequencing data is a key strength of the Bioconductor project. Starting with counts summarised at the gene-level, a typical analysis involves pre-processing, exploratory data analysis, differential expression testing and pathway analysis with the results obtained informing future experiments and validation studies. In this workflow article, we analyse RNA-sequencing data from the dorsal root ganglion neurons to unveil pain-related gene regulated by a human cellular tissue product. We demonstrate the use of the popular ***edgeR*** package to import, organise, filter and normalise the data, followed by the ***limma*** package with its voom method, linear modelling and empirical Bayes moderation to assess differential expression. This pipeline is further enhanced by minimizing the technical variations between each sets/batches by ComBat-seq package or the EdgeR package using the design formula (~batch + condition). 

##  Introduction
Gene expression is a highly regulated mechanism that controls the function and adaptability of all living cells including prokaryotes and eukaryotes. Several techniques exist for studying and quantifying gene expression and its regulation. Among various techniques deployed to assess gene expression, RNA-seq can provide qualitative (RNA sequence) and quantitative (RNA abundance) analyses of either targeted mRNA transcripts or the complete transcriptome of a particular issue. Transcriptomic data analysis can be performed using data obtained from a standard bulk RNA-seq. This tutorial describe the main steps to perform differential expression analysis (DEA). This is a brief introduction that covers the main aspects of DEA and outlines the most common types of analyses in this type of studies. The complete analysis also highlights the ease with which researchers can turn the raw counts from an RNA-sequencing experiment into biological insights using Bioconductor.

## Set-up

```{r}
# First, let's load all the packages we will need to analyse the data.
library(edgeR)
library(ggplot2)
library(Glimma)
library(gplots)
library(here)
library(limma)
library(NMF)
library(org.Mm.eg.db)
library(RColorBrewer)
library(statmod)
library(sva)
library(tidyverse)
```

## Data packaging

###  Reading in count-data
In the below, we first generate a ***DGEList-object*** from the count matrix.

```{r}
filename <- here()
filename
filename <- here(filename, "input/HHP_2h_24h_WT/")
```

```{r}
list.files(filename, recursive = TRUE)  # list all the files in this directory
```

Each of these text files contains the raw gene-level counts for a given sample. Note that our analysis only includes the Contol and HHP24h samples from this experiment (see associated file names below).

```{r}
sampleinfo <- read.delim(here(filename, "SampleInfo_without2h10min.txt"), stringsAsFactors = TRUE)
# sampleinfo <- read.delim('./SampleInfo_without2h10min.txt', stringsAsFactors
# = TRUE) # read for google colab
sampleinfo  # with treatment and batch information
```

```{r}
files <- c("Control_1_counts.txt", "Control_2_counts.txt", "Control_3_counts.txt",
    "24h_1_counts.txt", "24h_2_counts.txt", "24h_3_counts.txt")
# all the files
read.delim(here(filename, files[1]), nrow = 5)  # test first file
# read.delim(here(filename, './', files[1]), nrow = 5) # test first file counts
# file with geneid and counts
```

Whilst each of the six text files can be read into R separately and combined into a matrix of counts, edgeR offers a convenient way to do this in one step using the readDGE function. The resulting DGEList-object contains a matrix of counts with 35,976 rows associated with unique Entrez gene identifiers (IDs) and six columns associated with the individual samples in the experiment.

```{r}
x <- readDGE(here(filename, files), columns = c(1, 2))
# x <- readDGE(here(filename, './', files), columns = c(1, 2)) create DGE
# objects with 9 the counts files, only two columns
head(x)  # x with samples, counts, 
class(x)
dim(x)
names(x)
```

### Organising sample information 
For downstream analysis, sample-level information related to the experimental design needs to be associated with the columns of the counts matrix. This should include experimental variables, both biological and technical, that could have an effect on expression levels. Examples include cell type, genotype (wild-type, knock-out), phenotype (disease status, sex, age), sample treatment (drug, control) and batch information (date experiment was performed if samples were collected and analysed at distinct time points) to name just a few.

Our DGEList-object contains a samples data frame that stores both cell type (or group) and batch (sequencing lane) information, each of which consists of three distinct levels. Note that within x$samples, library sizes are automatically calculated for each sample and normalisation factors are set to 1. For simplicity, we remove the GEO sample IDs (GSM*) from the column names of our DGEList-object x.

```{r}
sampleinfo$SampleName
```

```{r}
colnames(x) <- sampleinfo$SampleName
```

```{r}
x$samples  # x, DGE objects with sample and counts dataframes
head(x$counts)
```

```{r}
group <- sampleinfo$Treatment
# Take a look
group
# Add the group information into the DGEList
x$samples$group <- group
x$samples
```

### Organising gene annotations
A second data frame named genes in the DGEList-object is used to store gene-level information associated with rows of the counts matrix. This information can be retrieved using organism specific packages such as Mus.musculus (Bioconductor Core Team 2016b) for mouse or the biomaRt package (Durinck et al. 2005, 2009) which interfaces the Ensembl genome databases in order to perform gene annotation.

The type of information that can be retrieved includes gene symbols, gene names, chromosome names and locations, Entrez gene IDs, Refseq gene IDs and Ensembl gene IDs to name just a few. biomaRt primarily works off Ensembl gene IDs, whereas Mus.musculus packages information from various sources and allows users to choose between many different gene IDs as the key.

The Entrez gene IDs available in our dataset were annotated using the Mus.musculus package to retrieve associated gene symbols and chromosome information.

```{r}
# Annotate the genes.
head(x$counts)
dim(x$counts)
geneid <- rownames(x)
genes <- AnnotationDbi::select(org.Mm.eg.db, keys = geneid, columns = c("ENTREZID",
    "SYMBOL", "GENENAME"), keytype = "SYMBOL")
# retrieve the ENTRZIS, GENENAME based on the SYMBOL keytype
head(genes)
genes <- genes[!duplicated(genes$ENTREZID), ]
```

As with any gene ID, Entrez gene IDs may not map one-to-one to the gene information of interest. It is important to check for duplicated gene IDs and to understand the source of duplication before resolving them.

```{r}
x <- x[genes$SYMBOL, ]  # keep only the genes with corresponding SYMBOLs in genes
head(x)
```

```{r}
x$genes <- genes  # add genes annotation to x
head(x)  # x has samples, counts, genes annotation attributes
table(genes$SYMBOL == rownames(x))  # test the genes and x rownames are the same
# 32879 genes total
```

In this example, the gene order is the same in both the annotation and the data object. If this is not the case due to missing and/or rearranged gene IDs, the match function can be used to order genes correctly. The data frame of gene annotations is then added to the data object and neatly packaged in a DGEList-object containing raw count data with associated sample information and gene annotations.

```{r}
# save the raw count maxtrix before batch correction
saveRDS(x, "./x_without_correction.RDS")
```

## Batch effect correction with ComBat-seq
Because the plotting of the gene loadings showed some batch effects, we minimized the technical variations between each batch during the differential expression analysis by using ComBat-seq package or the EdgeR package, which have been commonly used in previous studies.

ComBat-seq is a batch effect adjustment tool for bulk RNA-seq count data. It is an improved model based on the popular ComBat, to address its limitations through novel methods designed specifically for RNA-Seq studies. ComBat-seq takes untransformed, raw count matrix as input. Same as ComBat, it requires a known batch variable.

```{r}
# use ComBat-seq to correct batch effect, keep the counts
treatment = c("control", "control", "control", "hhp.24hr", "hhp.24hr", "hhp.24hr")
batch = c("first", "second", "third",  "first", "second", "third")
```

```{r}
groups = sapply(as.character(treatment), switch, control = 1, hhp.24hr = 2, USE.NAMES = F)
batches = sapply(as.character(batch), switch, first = 1, second = 2, third = 3, USE.NAMES = F)
# create levels to compare
```

```{r}
groups
batches
```

```{r}
x_corrected_counts <- ComBat_seq(counts = as.matrix(x), batch = batches, group = groups)
# transform counts
```

```{r}
head(x$counts)
head(x_corrected_counts)
```

```{r}
x$counts <- x_corrected_counts
```

```{r}
saveRDS(x, "./x_with_correction.RDS")
```

### Pre-filtering
While it is not necessary to pre-filter low count genes before running the functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within analysis. It can also improve visualizations, as features with no information for differential expression are not plotted.

Here we perform a minimal pre-filtering to keep only rows that have at least 10 reads total. Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.

```{r}
# Filtering lowly expressed genes Obtain CPMs
myCPM <- cpm(x)
# Have a look at the output
head(myCPM)
# Which values in myCPM are greater than 0.33?  A CPM of 0.33 is used as it
# corresponds to a count of 10-15 for the library sizes in this data set.  If
# the count is any smaller, it is considered to be very low, indicating that
# the associated gene is not expressed in that sample.  Let's have a look and
# see whether our threshold of 0.33 does indeed correspond to a count of about
# 10-15 We will look at the first sample
thresh <- myCPM > 0.33
# This produces a logical matrix with TRUEs and FALSEs
head(thresh)
# Summary of how many TRUEs there are in each row There are 11433 genes that
# have TRUEs in all 12 samples.
table(rowSums(thresh))
# we would like to keep genes that have at least 2 TRUES in each row of thresh
keep <- rowSums(thresh) >= 2
table(keep)
```

```{r}
plot(myCPM[, 1], x$counts[, 1])
# Let us limit the x and y-axis so we can actually look to see what is
# happening at the smaller counts
```

```{r}
plot(myCPM[, 1], x$counts[, 1], ylim = c(0, 50), xlim = c(0, 3))
# Add a vertical line at 0.5 CPM
abline(v = 0.33, h = 15, col = "blue")
# Now that we've checked our filtering method we will filter the DGEList
# object.)  Let us limit the x and y-axis so we can actually look to see what
# is happening at the smaller counts
```

```{r}
# Now that we've checked our filtering method we will filter the DGEList
# object.
x <- x[keep, keep.lib.sizes = FALSE]
```

```{r}
dim(x)
head(x)
```

### Normalising gene expression distributions
During the sample preparation or sequencing process, external factors that are not of biological interest can affect the expression of individual samples. For example, samples processed in the first batch of an experiment can have higher expression overall when compared to samples processed in a second batch. It is assumed that all samples should have a similar range and distribution of expression values. Normalisation is required to ensure that the expression distributions of each sample are similar across the entire experiment.

Any plot showing the per sample expression distributions, such as a density or boxplot, is useful in determining whether any samples are dissimilar to others. Distributions of log-CPM values are similar throughout all samples within this dataset (panel B of the figure above).

```{r}
# Quality control Library size and distribution plots First, we can check how
# many reads we have for each sample in the y.
x$samples$lib.size
# The names argument tells the barplot to use the sample names on the x-axis
# The las argument rotates the axis names
barplot(x$samples$lib.size, names = colnames(x), las = 2)
# Add a title to the plot
title("Barplot of library sizes")
```

```{r}
# we can also adjust the labelling if we want
barplot(x$samples$lib.size/1e+06, names = colnames(x), las = 2, ann = FALSE, cex.names = 0.75)
mtext(side = 1, text = "Samples", line = 0.5)
mtext(side = 2, text = "Library size (millions)", line = 3)
title("Barplot of library sizes")
```

```{r}
# Count data is not normally distributed, so if we want to examine the
# distributions of the raw counts we need to log the counts.  Get log2 counts
# per million
logcounts <- cpm(x, log = TRUE, prior.count = 1)
# Check distributions of samples using boxplots
boxplot(logcounts, xlab = "", ylab = "Log2 counts per million", las = 2)
# Let's add a blue horizontal line that corresponds to the median logCPM
abline(h = median(logcounts), col = "blue")
title("Boxplots of logCPMs (unnormalised)")
```

## Unsupervised clustering of samples
One of the most important exploratory plots to examine for gene expression analyses is the multi-dimensional scaling (MDS) plot, or similar. The plot shows similarities and dissimilarities between samples in an unsupervised manner so that one can have an idea of the extent to which differential expression can be detected before carrying out formal tests. Ideally, samples would cluster well within the primary condition of interest, and any sample straying far from its group could be identified and followed up for sources of error or extra variation. If present, technical replicates should lie very close to one another.

Such a plot can be made in limma using the plotMDS function. The first dimension represents the leading-fold-change that best separates samples and explains the largest proportion of variation in the data, with subsequent dimensions having a smaller effect and being orthogonal to the ones before it. When experimental design involves multiple factors, it is recommended that each factor is examined over several dimensions. If samples cluster by a given factor in any of these dimensions, it suggests that the factor contributes to expression differences and is worth including in the linear modelling. On the other hand, factors that show little or no effect may be left out of downstream analysis.

```{r}
# An MDSplot is a visualisation of a principle components analysis, which
# determines the greatest sources of variation in the data.
plotMDS(x)
```

```{r}
# Redo the MDSplot with corrected information
col.cell <- c("purple", "orange", "red", "blue")[sampleinfo$Treatment]
col.status <- c("#1B9E77", "#D95F02", "#7570B3")[sampleinfo$Batch]
char.celltype <- c(1, 4)[sampleinfo$Treatment]
plotMDS(x, col = col.cell)
legend("bottom", fill = c("purple", "orange", "red", "blue"), legend = levels(sampleinfo$Treatment))
title("Treatment")

p1 <- plotMDS(x, col = col.status)
legend("bottom", fill = c("#1B9E77", "#D95F02", "#7570B3"), legend = levels(sampleinfo$Batch),
    cex = 0.8)
title("Batch")
```

```{r}
char.status <- c(1, 3, 3)[sampleinfo$Batch]
plotMDS(x, dim = c(1, 2), col = col.cell, pch = char.status, cex = 1.5)
legend("bottom", legend = levels(sampleinfo$Treatment), col = c("purple", "orange",
    "red", "blue"), pch = 16)
legend("right", legend = levels(sampleinfo$Batch), pch = c(16, 3, 4))
```

### Hierarchical clustering with heatmaps
A heatmap is created for the top 200 most variable genes and plot the heatmap using the heatmap.2 function from the gplots package. The heatmap correctly clusters samples by cell type and reorders the genes into blocks with similar expression patterns. From the heatmap, we observe that the expression of contol and drug samples are very similar for the top 200 most variable genes.

```{r}
# Let's select data for the 500 most variable genes and plot the heatmap We
# estimate the variance for each row in the logcounts matrix
var_genes <- apply(logcounts, 1, var)
head(var_genes)
# Get the gene names for the top 500 most variable genes
select_var <- names(sort(var_genes, decreasing = TRUE))[1:200]
head(select_var)
# Subset logcounts matrix
highly_variable_lcpm <- logcounts[select_var, ]
dim(highly_variable_lcpm)
head(highly_variable_lcpm)
```

```{r}
## Get some nicer colours
mypalette <- brewer.pal(11, "RdYlBu")
morecols <- colorRampPalette(mypalette)
# Set up colour vector for celltype variable
col.status <- c("#1B9E77", "#D95F02", "#7570B3")[sampleinfo$Batch]
# Plot the heatmap
heatmap.2(highly_variable_lcpm, scale = "row", col = rev(morecols(100)), density.info='none', Colv = F, trace = "none",
    main = "Top 200 most variable genes across samples", ColSideColors = col.status)
```

### Normalisation for composition bias

```{r}
# The calcNormFactors function calculates the normalization factors between
# libraries.  TMM normalisation (and most scaling normalisation methods) scale
# relative to one sample.  Apply normalisation to DGEList object
x <- calcNormFactors(x)
x$samples
# The last two samples have much smaller normalisation factors, and MCL1.LA and
# MCL1.LB have the largest.  If we plot mean difference plots using the plotMD
# function for these samples, we should be able to see the composition bias
# problem.
par(mfrow = c(1, 2))
plotMD(logcounts, column = 1)
abline(h = 0, col = "grey")
plotMD(logcounts, column = 4)
abline(h = 0, col = "grey")
# if we redo these plots using y, we should see the composition bias problem
# has been solved.
par(mfrow = c(1, 2))
plotMD(x, column = 1)
abline(h = 0, col = "grey")
plotMD(x, column = 4)
abline(h = 0, col = "grey")
# Plot the biased and unbiased MD plots side by side for the same sample to see
# the before and after TMM normalisation effect.
par(mfrow = c(1, 2))
plotMD(logcounts, column = 3, main = "Before TMM normalisation")
abline(h = 0, col = "grey")
plotMD(x, column = 3, main = "After TMM normalisation")
abline(h = 0, col = "grey")
```

```{r}
x[select_var, ]
```

```{r}
saveRDS(x, "./x_with_correction.rds")
# saveRDS(x, './x_without_correction.rds')
```

## Differential expression with limma-voom
Linear modelling in limma is carried out using the lmFit and contrasts.fit functions originally written for application to microarrays. The functions can be used for both microarray and RNA-seq data and fit a separate model to the expression values for each gene. Next, empirical Bayes moderation is carried out by borrowing information across all the genes to obtain more precise estimates of gene-wise variability. The model’s residual variances are plotted against average expression values in the next figure. It can be seen from this plot that the variance is no longer dependent on the mean expression level.

```{r}
x <- readRDS("./x_with_correction.rds")
summary(x)
```

```{r}
# Create the design matrix Look at group variable again
treatment <- sampleinfo$Treatment
# Take a look
treatment
# Convert to factor
treatment <- factor(treatment)
# Specify a design matrix without an intercept term
design <- model.matrix(~0 + treatment)
design
# Make the column names of the design matrix a bit nicer
colnames(design) <- levels(treatment)
design
```

```{r}
# Voom transform the data
v <- voom(x, design, plot = TRUE)
head(v)
# What is contained in this object?
names(v)
dim(v$weights)
# We can repeat the box plots for the normalised data to compare to before
# normalisation.  The expression values in v$E are already log2 values so we
# don't need to log-transform.
par(mfrow = c(1, 2))
boxplot(logcounts, xlab = "", ylab = "Log2 counts per million", las = 2, main = "Unnormalised logCPM")
## Let's add a blue horizontal line that corresponds to the median logCPM
abline(h = median(logcounts), col = "blue")
boxplot(v$E, xlab = "", ylab = "Log2 counts per million", las = 2, main = "Voom transformed logCPM")
## Let's add a blue horizontal line that corresponds to the median logCPM
abline(h = median(v$E), col = "blue")
```

## Testing for differential expression, contro vs HHP.24h

```{r}
# Now that we have the voom transformed data we can use limma to test for
# differential expression.  Fit the linear model
fit <- lmFit(v)
names(fit)
# The comparison of interest can be specified using the makeContrasts function.
cont.matrix <- makeContrasts(convs24h = hhp.24hr - control, levels = design)
cont.matrix
# Now we can apply the contrasts matrix to the fit object to get the statistics
# and estimated parameters of our comparison that we are interested in.  Now we
# can apply the contrasts matrix to the fit object to get the statistics and
# estimated parameters of our comparison that we are interested in.
fit.cont.B <- contrasts.fit(fit, cont.matrix)
# The final step is to call the eBayes function, which performs empirical Bayes
# shrinkage on the variances, and estimates moderated t-statistics and the
# associated p-values.
fit.cont.B <- eBayes(fit.cont.B)
# Check the dimensions of the fit object
dim(fit.cont.B)
# We can use the limma decideTests function to generate a quick summary of DE
# genes for the contrasts.
summa.fit.B <- decideTests(fit.cont.B, adjust.method = "BH", lfc = 0.5)
summary(summa.fit.B)
```

```{r}
head(fit.cont.B)
```

```{r}
seq_tab <- topTable(fit.cont.B, adjust = "BH", n = "inf")
head(seq_tab)
dim(seq_tab)
```

```{r}
write.csv(seq_tab, "RNA-seqResults-Convs24h.csv")
```

```{r}
saveRDS(seq_tab, "./seq_tab.rds")
```

### Plots after testing for DE
To summarise results for all genes visually, mean-difference plots, which display log-FCs from the linear model fit against the average log-CPM values can be generated using the plotMD function, with the differentially expressed genes highlighted.

```{r}
# We want to highlight the significant genes. We can get this from decideTests.
plotMD(fit.cont.B, coef = 1, status = summa.fit.B[, "convs24h"], values = c(-1, 1),
    hl.col = c("blue", "red"))
# For the volcano plot we have to specify how many of the top genes to
# highlight.  We can also specify that we want to plot the gene symbol for the
# highlighted genes.  let's highlight the top 100 most DE genes
volcanoplot(fit.cont.B, coef = 1,  names = fit.cont.B$genes$SYMBOL,
    main = "convs24h")
# ook at the MD plot and volcano plot for the second comparison, L.PregVsLac.
# Change the number of highlighted genes to 200 in the volcano plot.
```

### Examining the number of DE genes
For a quick look at differential expression levels, the number of significantly up- and down-regulated genes can be summarised in a table. Significance is defined using an adjusted p-value cutoff that is set at 5% by default. For the comparison between expression levels in basal and LP, 4,648 genes are found to be down-regulated in basal relative to LP and 4,863 genes are up-regulated in basal relative to LP – a total of 9,511 DE genes. A total of 9,598 DE genes are found between basal and ML (4,927 down- and 4,671 up-regulated genes), and a total of 5,652 DE genes are found between LP and ML (3,135 down- and 2,517 up-regulated). The larger numbers of DE genes observed for comparisons involving the basal population are consistent with our observations from the MDS plots.

```{r}
summary(summa.fit.B)
```

### Testing relative to a threshold (TREAT)
In other words, you are not controlling the false discovery rate at 5% any  more. There is a function called treat in the limma package that performs this style of analysis correctly (McCarthy and Smyth 2009).  Let's decide that we are only interested in genes that have a absolute logFC of 1.  This corresponds to a fold change of 2, or 0.5 (i.e.  double or half).  We can perform a treat analysis which ranks our genes according to p-value AND logFC.  This is easy to do after our analysis, we just give the treat function the fit.cont object and specify our cut-off.

```{r}
fit.treat.B <- treat(fit.cont.B, lfc = 0.3)
res.treat.B <- decideTests(fit.treat.B)
summary(res.treat.B)

# Calculate the treat seperately
topTable(fit.treat.B, coef = 1, sort.by = "p")
# Notice that much fewer genes are highlighted in the MAplot
par(mfrow = c(1, 2))
plotMD(fit.treat.B, coef = 1, status = res.treat.B[, "convs24h"], values = c(-1,
    1), hl.col = c("blue", "red"))
abline(h = 0, col = "grey")

# An interactive version of the mean-difference plots is possible via the
# glMDPlot function in the Glimma package.
glMDPlot(fit.treat.B, coef = 1, counts = v$E, groups = treatment, status = res.treat.B,
    side.main = "ENTREZID", main = "convs24h", folder = "md")
```

```{r}
topTable(fit.treat.B, coef = 1, sort.by = "p")
```

## Select gene for GO analysis
Further analyses are required to interpret the differential expression results in a biological context. One common downstream procedure is a gene ontology (GO) enrichment analysis.

The analysis of significantly enriched GO (Gene Onthology) terms over-represented among DEG lists was performed with the ClusterProfiler package in R later. We need select the DEG for further GO analysis.

```{r}
# select the up-regulated genes, summa.fit.B == 1 (down-regulated, summa.fit.B == -1)
select_genes <- summa.fit.B == 1
sum(select_genes)
head(select_genes)
```

```{r}
head(fit.cont.B[select_genes, ])
```

```{r}
forGO <- cbind(fit.cont.B[select_genes, ]$coefficients, fit.cont.B[select_genes,
    ]$p.value)
# choose the gene for further GO analysis differential genes after ComBat-seq
# correction
dim(forGO)
```

```{r}
write.csv(forGO, "forGO_up.csv")
```

## Differential expression after batch effect correction with edgeR
We also can use the mixed model from edgeR for batch-effect correction.
In this case, additive model formulas are used is when correcting for batch effects in an experiment. The situation here is analogous to blocking, the only difference being that the batch effects were probably unintended rather than a deliberate aspect of the experimental design. The analysis is the same as for blocking. The treatments can be adjusted for differences between the batches by using an additive model formula of the form.

```{r}
x <- readRDS("./x_without_correction.rds")
summary(x)
```

```{r}
# Create the design matrix Look at group variable again
treatment <- sampleinfo$Treatment
treatment <- factor(treatment)
batch <- sampleinfo$Batch
batch <- factor(batch)
treatment
batch
```

```{r}
# Specify a design matrix without an intercept term Make the column names of
# the design matrix a bit nicer
design <- model.matrix(~batch + treatment)
rownames(design) <- colnames(x)
design
```

```{r}
x <- estimateDisp(x, design, robust = TRUE)
x$common.dispersion
plotBCV(x)
```

```{r}
fit <- glmQLFit(x, design, robust = TRUE)
plotQLDisp(fit)
```

```{r}
qlf <- glmQLFTest(fit, coef = 1:4)
topTags(qlf)
```

```{r}
FDR <- p.adjust(qlf$table$PValue, method = "BH")
sum(FDR < 0.05)
```

```{r}
qlf <- glmQLFTest(fit, coef = 4)
topTags(qlf)
```

```{r}
summa.fit.L <- decideTests(qlf, lfc = 0.5)
summary(summa.fit.L)
```

```{r}
top <- rownames(topTags(qlf))
cpm(x)[top, ]
```

```{r}
plotMD(qlf)
abline(h = c(-1, 1), col = "blue")
```

## Examining the number of DE genes identified after batch effect correction.
We can also examine the number of DE genes identified after batch effect correction by ComBat-seq, edgeR.
The Venn diagram showed a substantial overlap between the DEGs following batch effect correction by using either ComBat-seq or edgeR, suggesting that this approach is consistent and reliable to adjust the batch effect.

```{r}
summary(summa.fit.B)
summary(summa.fit.L)
```

```{r}
summa.fit.B <- rownames_to_column(as.data.frame(summa.fit.B)) %>%
    rename(ComBat_seq = "convs24h")
summa.fit.L <- rownames_to_column(as.data.frame(summa.fit.L)) %>%
    rename(edgeR = "treatmenthhp.24hr")
c3 <- left_join(summa.fit.B, summa.fit.L, by = "rowname")
c3 <- column_to_rownames(c3, var = "rowname")
head(c3)
dim(c3)
```

```{r}
# Add another contrast to the contrasts matrix, check out the vennDiagram
# function c3 <- cbind(summa.fit.B, summa.fit.L)
head(c3)
vennDiagram(c3, include = c("up", "down"), names = c("ComBat_seq", "edgeR"), counts.col = c("red",
    "blue"), circle.col = c("turquoise", "salmon"))
```

## References
Michael I. Love, Simon Anders, and Wolfgang Huber. 2022. “Analyzing RNA-seq data with DESeq2.” Bioconductor, June. http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html.

Charity W. Law, Monther Alhamdoosh, Shian Su, Xueyi Dong1, Luyi Tian, Gordon K. Smyth, Matthew E. Ritchie. 2016. “RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR.” F1000Res May. https://f1000research.com/articles/5-1408.

Juan R Gonzalez. 2021. “Transciptomic analysis using limma and limma + voom pipelines.” RPubs by RStudio, May. https://rpubs.com/jrgonzalezISGlobal/transcriptomic_analyses.

## Session Info

```{r}
sessionInfo()
```

